pageserver/src/page_service.rs

```diff
@@ -201,7 +201,7 @@ pub fn thread_main(
             .name("serving Page Service thread".into())
             .spawn(move || {
                 if let Err(err) = page_service_conn_main(conf, local_auth, socket, auth_type) {
-                    error!(%err, "page server thread exited with error");
+                    error!("page server thread exited with error: {:?}");
```

Comment:
```suggestion
                    error!("page server thread exited with error: {:?}", err);
```

Reply:
> One question (not necessary requires fixing): What happens if we omit `RUST_BACKTRACE` and print such errors? Before, it did not matter and it printed the whole context, would be nice to have it still (or ensure that `RUST_BACKTRACE` is always set?)

We are currently always setting RUST_BACKTRACE=1 when the pageserver is launched by the `zenith` cli. See `fill_rust_env_vars` in `control_plane/src/lib.rs`. With that, an error in the log looks like this:
```
2022-01-13T20:16:24.119288Z ERROR Error processing HTTP request: InternalServerError(invalid branch start lsn

Caused by:
    LSN 0/48 is earlier than initdb lsn 0/1698888

Stack backtrace:
   0: <pageserver::layered_repository::LayeredTimeline as pageserver::repository::Timeline>::check_lsn_is_in_scope
   1: <pageserver::layered_repository::LayeredRepository as pageserver::repository::Repository>::branch_timeline
             at /home/heikki/git-sandbox/zenith/pageserver/src/layered_repository.rs:206:9
   2: pageserver::branches::create_branch
             at /home/heikki/git-sandbox/zenith/pageserver/src/branches.rs:340:5
   3: pageserver::http::routes::branch_create_handler::{{closure}}::{{closure}}
             at /home/heikki/git-sandbox/zenith/pageserver/src/http/routes.rs:81:9
      <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll
             at /home/heikki/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.12.0/src/runtime/blocking/task.rs:42:21
   4: tokio::runtime::task::core::CoreStage<T>::poll::{{closure}}
             at /home/heikki/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.12.0/src/runtime/task/core.rs:161:17
      tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /home/heikki/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.12.0/src/loom/std/unsafe_cell.rs:14:9
   5: tokio::runtime::task::core::CoreStage<T>::poll
             at /home/heikki/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.12.0/src/runtime/task/core.rs:151:13
      tokio::runtime::task::harness::poll_future::{{closure}}
             at /home/heikki/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.12.0/src/runtime/task/harness.rs:461:19
      <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /usr/src/rustc-1.56.0/library/core/src/panic/unwind_safe.rs:271:9
      std::panicking::try::do_call
             at /usr/src/rustc-1.56.0/library/std/src/panicking.rs:403:40
      std::panicking::try
             at /usr/src/rustc-1.56.0/library/std/src/panicking.rs:367:19
      std::panic::catch_unwind
             at /usr/src/rustc-1.56.0/library/std/src/panic.rs:129:14
      tokio::runtime::task::harness::poll_future
             at /home/heikki/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.12.0/src/runtime/task/harness.rs:449:18
      tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at /home/heikki/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.12.0/src/runtime/task/harness.rs:98:27
      tokio::runtime::task::harness::Harness<T,S>::poll
             at /home/heikki/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.12.0/src/runtime/task/harness.rs:53:15
   6: tokio::runtime::task::raw::RawTask::poll
             at /home/heikki/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.12.0/src/runtime/task/raw.rs:70:18
      tokio::runtime::task::UnownedTask<S>::run
             at /home/heikki/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.12.0/src/runtime/task/mod.rs:379:9
      tokio::runtime::blocking::pool::Inner::run
             at /home/heikki/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.12.0/src/runtime/blocking/pool.rs:265:17
   7: tokio::runtime::blocking::pool::Spawner::spawn_thread::{{closure}}
             at /home/heikki/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.12.0/src/runtime/blocking/pool.rs:245:17
      std::sys_common::backtrace::__rust_begin_short_backtrace
             at /usr/src/rustc-1.56.0/library/std/src/sys_common/backtrace.rs:125:18
   8: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}
             at /usr/src/rustc-1.56.0/library/std/src/thread/mod.rs:481:17
      <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /usr/src/rustc-1.56.0/library/core/src/panic/unwind_safe.rs:271:9
      std::panicking::try::do_call
             at /usr/src/rustc-1.56.0/library/std/src/panicking.rs:403:40
      std::panicking::try
             at /usr/src/rustc-1.56.0/library/std/src/panicking.rs:367:19
      std::panic::catch_unwind
             at /usr/src/rustc-1.56.0/library/std/src/panic.rs:129:14
      std::thread::Builder::spawn_unchecked::{{closure}}
             at /usr/src/rustc-1.56.0/library/std/src/thread/mod.rs:480:30
      core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /usr/src/rustc-1.56.0/library/core/src/ops/function.rs:227:5
   9: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /usr/src/rustc-1.56.0/library/alloc/src/boxed.rs:1636:9
      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /usr/src/rustc-1.56.0/library/alloc/src/boxed.rs:1636:9
      std::sys::unix::thread::Thread::new::thread_start
             at /usr/src/rustc-1.56.0/library/std/src/sys/unix/thread.rs:106:17
  10: start_thread
             at ./nptl/./nptl/pthread_create.c:481:8
  11: __clone
             at ./misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:95)
```

If you remove that RUST_BACKTRACE=1 from `fill_rust_env_vars`, it looks like this:
```
2022-01-13T22:00:35.678765Z ERROR Error processing HTTP request: InternalServerError(invalid branch start lsn

Caused by:
    LSN 0/48 is earlier than initdb lsn 0/1698888)
```

@cicdteam, in staging/production, do we currently use the `zenith` CLI to launch the page server? If not, do we set RUST_BACKTRACE=1? It would be good to do that.

Reply:
Good, so we have the `context` printed with `{:?}` similar to `{:#}` even without backtraces.

Reply:
@hlinnaka 

pageserver managed by systems an there is necessary setting:

```
Environment=RUST_BACKTRACE=1 ZENITH_REPO_DIR=/storage/pageserver LD_LIBRARY_PATH=/usr/local/lib
ExecStart=/usr/local/bin/pageserver -c "pg_distrib_dir='/usr/local'" -c "listen_pg_addr='0.0.0.0:6400'" -c "listen_http_addr='0.0.0.0:9898'" -D /storage/pageserver/data
```